### PR TITLE
feat: Update time-related sorters and move them into `Category` filter

### LINKED ISSUE #1871: [Feature] Improve “Recently Launched” filter
### What projects is this feature for?

OpenChallenges

### Description

One of the sorter for challenges is "Recently launched". The current implementation takes all the challenges whose start date is in the past and sort them so that the closer the start date is from today, the higher it will rank in the results.

There are a few issues with this approach:

- the results may include challenges that have started a long time ago, so they are not really "recently" launched
- the results may include closed challenges

One improvement would be to rename the sorter to "Start date (most recently)" to make it clearer that the sorting is done based on the start date. But shall we keep displaying only challenges that have already started?

Another options would be to create a filter that shows only all challenges that have started recently, which may require use to define what "recently" means (e.g. within the past 3 months).

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by rrchai**
- closes #1871 

Notes from the standup meeting:
As @vpchung and @tschaffter pointed out, some challenges only take 1 month or so to complete, but we also would like to show as many (reasonable) challenges as possible.

### ChangeLogs:

~~Update time-related sorters within a fixed window:~~
- ~~`Recently Launched`: start date from 3 months ago to now~~
- ~~`Recently Completed`: end date from 3 months ago to now~~
- ~~`Starting Soon`: start date from now to next 3 months~~
- ~~`Closing Soon`: end date from now to next 3 months~~

- Convert time-related sorters into `Category` filter using "3 months window + filter corresponding status":
  - https://github.com/rrchai/sage-monorepo/pull/3
  - `Recently Launched`: start date from 3 months ago to now + status is "active" 
  - `Recently Completed`: end date from 3 months ago to now + status is "completed"
  - `Starting Soon`: start date from now to next 3 months + status is "upcoming" 
  - `Closing Soon`: end date from now to next 3 months + status is "active"

- Add ~~"Newest"~~ "Start Date" sorter 
  - https://github.com/rrchai/sage-monorepo/pull/4
  - Use only ~~"Newest"~~ "Start Date" sorter that sorts challenges by `startDate` (newest to oldest by default) - [691bca](https://github.com/Sage-Bionetworks/sage-monorepo/pull/2163/commits/691bca9d8222d0e6b095e34691fbe5888250689c)

- Move up `Category` filter

#### ~~"Newest"~~ "Start Date" sorter

https://github.com/Sage-Bionetworks/sage-monorepo/assets/73901500/ece7fa5a-74c0-431b-8a4a-305024d2a79e

#### New values in `Category` filter

https://github.com/Sage-Bionetworks/sage-monorepo/assets/73901500/26d0442e-9919-4662-a6b0-1f1dbf3ba6c4

### CODE REVIEW

**Comment by rrchai** [comment]
## Use _three months_ as the time window:

#### Recently Launched
Screen Shot 2023-09-27 at 2 52 14 PM

#### Recently Completed
Screen Shot 2023-09-27 at 2 52 28 PM

#### Starting Soon
Screen Shot 2023-09-27 at 2 52 52 PM

#### Closing Soon
Screen Shot 2023-09-27 at 2 53 00 PM

**Comment by rrchai** [comment]
Remaining issues with three months window alone:
- The active challenges in "Recently Completed" results has passed the end dates already.
- There might be confusions for users when the upcoming challenges are in the "Closing Soon" results.
- Some users might expect that "sort" will not filter the data, as what has been currently processed in the backend.

**Comment by rrchai** [comment]
> - The status of challenge is mismatched the start/end date of challenge. For example, the active challenges in "Recently Completed" results has passed the end dates already.
> - There might be confusions for users when the upcoming challenges are also in the "Closing Soon" results.

1. One possible solution is to also add status filters in addition to start/end filter:
- Recently Launched: start date from 3 months ago to now + status is "active"
- Recently Completed: end date from 3 months ago to now + status is "completed"
- Starting Soon: start date from now to next 3 months + status is "upcoming" 
- Closing Soon: end date from now to next 3 months + status is "active"

pros: challenge results are expected for most users regardless of whether the status of challenge is outdated.
cons: reduce a few results

2. Or we could ensure the status and start/end always matched, and try smaller time window 
pros: preserve most results
cons: there might be still some upcoming challenges in the "Closing Soon" (i.e. we have a challenge start: Oct 19, end: Oct 20 - hmm interesting 😄 )

**Comment by rrchai** [comment]
```console
GET {{basePath}}/challenges?sort=ending_soon

{
  "number": 0,
  "size": 100,
  "totalElements": 1,
  "totalPages": 1,
  "hasNext": false,
  "hasPrevious": false,
  "challenges": [
    {
      "id": 177,
      "slug": "rsna-2023-abdominal-trauma-detection",
      "name": "RSNA 2023 Abdominal Trauma Detection",
      "startDate": "2023-07-26",
      "endDate": "2023-10-13",
      "status": "active",
      ...
  ]
}
```

**Comment by tschaffter** [comment]
> Move these four sorters into the Category filter

Could you estimate the complexity of taking this road? From a UI perspective it makes the most sense. Also we could then show at the same time "challenges that started recently" and "challenges that are starting soon", for example. Then the user can apply the sorter of their choice, e.g. by start date (new sorter), end date (new sorter), most stared, etc.

**Comment by rrchai** [comment]
> Could you estimate the complexity of taking this road?

well, I believe these are steps:
1. update backend to transfer these 4 sorters to filters in `ChallengeSearchQuery`
2. update frontend to convert sorters into values of Category filter 

I would say it may require half day for me to do so.

> From a UI perspective it makes the most sense. Also we could then show at the same time "challenges that started recently" and "challenges that are starting soon", for example. Then the user can apply the sorter of their choice, e.g. by start date (new sorter), end date (new sorter), most stared, etc.

I like the idea. I can go for this approach tomorrow and try to finish early afternoon, so that we have time to review it.

**Comment by rrchai** [comment]
Done, the preview is updated with latest version.

**Comment by rrchai** [comment]
I initially added both `Start Date` and `End Date` sorters, but I changed to only add `Newest` sorter (based on `start_date`) in UI for two reasons:
- It's kind of hard to find user friendly names when both "end date" and "start date" sorters are used at the same time
- I don't see "Oldest" (sort by oldest -> newest) is the sorter that will be frequently used. In addition, "Oldest" might be slightly have negative affect?

(the backend should be still able to sort by both `start_date` and `end_date`, so we have more flexibility to add back later if needed)

---
The preview has been updated with the latest version.

**Comment by rrchai** [comment]
rename "newest" to "start date"
reorder filters "incentive" -> "status" -> "categories"

File: apps/openchallenges/challenge-service/requests.http [hunk 1]
```diff
@@ -52,30 +52,37 @@ GET {{basePath}}/challenges?minStartDate=2017-01-01&maxStartDate=2017-12-31
...
 GET {{basePath}}/challenges?inputDataTypes=genomic
 
-
-### List the challenge that categorized as featured
+### List the challenge that is categorized as featured
 
 GET {{basePath}}/challenges?categories=featured
 
-### Sort the challenges by their starred count in desc order
+### List the challenges that is categorized as starting soon .
```
Comment (tschaffter) [inline apps/openchallenges/challenge-service/requests.http:59]:
```suggestion
### List the challenges that are categorized as starting soon.
```

File: apps/openchallenges/challenge-service/requests.http [hunk 2]
```diff
@@ -52,30 +52,37 @@ GET {{basePath}}/challenges?minStartDate=2017-01-01&maxStartDate=2017-12-31
...
 
-GET {{basePath}}/challenges?sort=starred&direction=desc
+GET {{basePath}}/challenges?categories=starting_soon
+
+### List the challenges that is categorized as ending soon.
```
Comment (tschaffter) [inline apps/openchallenges/challenge-service/requests.http:63]:
```suggestion
### List the challenges that are categorized as ending soon.
```

File: apps/openchallenges/challenge-service/requests.http [hunk 3]
```diff
@@ -52,30 +52,37 @@ GET {{basePath}}/challenges?minStartDate=2017-01-01&maxStartDate=2017-12-31
...
+
+GET {{basePath}}/challenges?categories=ending_soon
 
-### Sort the challenges by how soon they start.
+### List the challenges that is categorized as recently started.
```
Comment (tschaffter) [inline apps/openchallenges/challenge-service/requests.http:67]:
```suggestion
### List the challenges that are categorized as recently started.
```

File: apps/openchallenges/challenge-service/requests.http [hunk 4]
```diff
@@ -52,30 +52,37 @@ GET {{basePath}}/challenges?minStartDate=2017-01-01&maxStartDate=2017-12-31
...
 
-GET {{basePath}}/challenges?sort=starting_soon
+GET {{basePath}}/challenges?categories=recently_started
 
-### Sort the challenges by how soon they end.
+### List the challenges that is categorized as recently ended .
```
Comment (tschaffter) [inline apps/openchallenges/challenge-service/requests.http:71]:
```suggestion
### List the challenges that are categorized as recently ended.
```

File: apps/openchallenges/challenge-service/src/main/java/org/sagebionetworks/openchallenges/challenge/service/model/repository/CustomChallengeRepositoryImpl.java [hunk 5]
```diff
@@ -256,10 +256,51 @@ private SearchPredicate getChallengeIncentivesPredicate(
    */
   private SearchPredicate getCategoriesPredicate(
       SearchPredicateFactory pf, ChallengeSearchQueryDto query) {
+
+    LocalDate now = LocalDate.now();
+    LocalDate lastQuarter = now.minusMonths(3);
```
Comment (tschaffter) [inline apps/openchallenges/challenge-service/src/main/java/org/sagebionetworks/openchallenges/challenge/service/model/repository/CustomChallengeRepositoryImpl.java:261]: > A quarter is a three-month period on a company's financial calendar that acts as a basis for periodic financial reports and the paying of dividends.

The variable name may introduce confusion as quarter are not any three-month periods but correspond to specific months (e.g. Q1 is often Jan-March).

Maybe `recentPeriod` and `soonPeriod`

**Review by tschaffter — CHANGES_REQUESTED** [review]

File: apps/openchallenges/challenge-service/src/main/java/org/sagebionetworks/openchallenges/challenge/service/model/repository/CustomChallengeRepositoryImpl.java (hunk 5 above)
Comment (rrchai) [reply apps/openchallenges/challenge-service/src/main/java/org/sagebionetworks/openchallenges/challenge/service/model/repository/CustomChallengeRepositoryImpl.java:261]: I agree there will be confusion. 
I changed to use "threeMonthsAgo" and "threeMonthsLater" to be more specific and clear.

**Review by tschaffter — APPROVED** [review]

**Comment by vpchung** [comment]
This is great, thanks!

> we also would like to show as many (reasonable) challenges as possible.

This makes sense, but I still think 3 months is a bit long as a threshold for "Starting Soon" and "Closing Soon".  My concern is that filtering by "Starting Soon" and "Upcoming" would return the same results, so what value is there to have the two filters?

**Comment by rrchai** [comment]
> This makes sense, but I still think 3 months is a bit long as a threshold for "Starting Soon" and "Closing Soon". My concern is that filtering by "Starting Soon" or "Upcoming" would return the same results, so what value is there to have the two filters?

Thank you for bringing this up. This is a good point to consider to  prevent filter overlaps. Currently, we do have challenges with launches scheduled for more than three months. However, no new upcoming challenges are added, we can anticipate an increasing overlap between the "Starting Soon" and "Upcoming" categories moving forward.

Screen Shot 2023-09-28 at 2 11 50 PM
Screen Shot 2023-09-28 at 2 12 03 PM

--- 
So how about narrow the window for "Closing Soon" and "Starting Soon" to 1.5 or 2 month as a middle point, as well as more clear distinguish between status filter? @tschaffter  @vpchung

**Comment by tschaffter** [comment]
> So how about narrowing the window for "Closing Soon" and "Starting Soon" to ONE month

I agree with this. To clarify, "recently" would remain as is (3 months ago), right?

**Comment by rrchai** [comment]
> To clarify, "recently" would remain as is (3 months ago), right?

yep

**Comment by vpchung** [comment]
> So how about narrowing the window for "Closing Soon" and "Starting Soon" to ONE month?

One month sounds very reasonable to me.

**Pull request merged**